### ISSUE TITLE
Inconsistent typescript version causes plugin compile error

### ISSUE THREAD

**Issue by pilaoda** [TypeScriptToLua/TypeScriptToLua#1132]
```
const plugin: tstl.Plugin = {
    visitors: {
        // Visitor can be a function that returns Lua AST node
        [ts.SyntaxKind.SourceFile]: {} as any,
    },
};
```

> Type '{ 300: any; }' is not assignable to type 'Visitors'.
  Object literal may only specify known properties, and '[ts.SyntaxKind.SourceFile]' does not exist in type 'Visitors'.

This may be caused by inconsistend typescript version inside tstl's node_module and my own node_module.
![image](https://user-images.githubusercontent.com/12658348/132536386-793e69bf-8c24-444a-bd89-0a559826bdf6.png)

![image](https://user-images.githubusercontent.com/12658348/132536636-95bd0945-d302-4145-bee8-17ebe7476fca.png)
![image](https://user-images.githubusercontent.com/12658348/132536732-ab767fa2-d976-4198-9cec-b5a007332bbb.png)

The same enum has different value! 
Why tstl need typescript inside its node_module? 
I just run `npm install -D typescript-to-lua` in my main folder.

My own typescript version is 4.4.2, and tstl has the 4.3.5 version.

**Comment by Perryvw** [comment]
tstl depends directly on TS 4.3 because we use the TypeScript compiler API, and that API changes between versions like you have discovered. You indeed need to make sure you have the same version installed in your project, otherwise it might not work well. Maybe we could make this a peer dependency instead, but I'm not entirely sure that would help.

**Comment by thejustinwalsh** [comment]
You must use the same version of typescript that tstl depends on.
I have run into this issue too.

The `package.json` for typescript within the `typescript-to-lua` package is defined as `"typescript": "~4.3.5"` though I have had issues even with a minor version bump, and am suggesting strict equality. in the package.json.

Yarn and nom both support mechanisms to force resolutions as well, and you may want to follow this practice in your own projects as well.

see:
https://www.npmjs.com/package/npm-force-resolutions
https://classic.yarnpkg.com/en/docs/selective-version-resolutions/

If you did not declare any typescript version typescript would be brought in with the correct version from it's dependencies.
Additionally if you are working on a plugin or dependency that needs to co-exist with typescript-to-lua you may want to consider peer dependencies as well.

```json
"peerDependencies": {
    "typescript": "~4.3.5",
    "typescript-to-lua": "^1.0.0"
  },
```

TLDR you do not want two different versions of typescript showing up in your yarn.lock or package lock.json.
Outside of the specifics of TypeScriptToLua, depending on multiple versions of a package can greatly increase your bundle size as well, so I would argue that it is a best practice to not find yourself with multiple versions of a dependency unless strictly needed.

**Comment by pilaoda** [comment]
Got it. Uninstall typescript 4.4.2 and reinstall tstl with auto typescript 4.3.5 solve my problem.
